### ISSUE TITLE
OutOfMemoryError when making multiple silent auth async calls | Stress Test

### ISSUE THREAD

**Issue by heidijinxujia** [AzureAD/microsoft-authentication-library-for-android#145]
*Stress Test Details*
    Open 10 threads calling async acquireTokenSilient at the same time for 8 hours, while there is RT in the cache.

*Problem*
    After 3 hours, got the OutOfMemory error. I could not use the semaphore here, because the mHandler is queuing all the msg which prevent the calls to hit the callback after getting the auth result. Without hitting the callback, the permit in the semaphore cannot be released.

```
04-27 22:47:54.349 7339-7633/com.microsoft.identity.client.sample I/art: Alloc concurrent mark sweep GC freed 886(30KB) AllocSpace objects, 0(0B) LOS objects, 3% free, 443MB/459MB, paused 2.285ms total 657.607ms
04-27 22:47:54.349 7339-7351/com.microsoft.identity.client.sample I/art: WaitForGcToComplete blocked for 1.446s for cause HeapTrim
04-27 22:47:54.349 7339-7339/com.microsoft.identity.client.sample W/System.err: WTF Thread: 20
04-27 22:47:54.351 7339-7633/com.microsoft.identity.client.sample E/art: Throwing OutOfMemoryError "Failed to allocate a 81826734 byte allocation with 16777120 free bytes and 68MB until OOM"
```

*Code*
```
private void callCacheStressTest(final String[] scopes, final User user,  final boolean forceRefresh, final boolean clearCache) throws InterruptedException{

	final int MAX_AVAILABLE = 10;
	final int TIME_LIMIT = 8 * 60 * 60 * 1000;
	final long startTime = System.currentTimeMillis();
	final ExecutorService executor = Executors.newFixedThreadPool(MAX_AVAILABLE);

	//clear the cache
	if(clearCache && user != null) {
		mApplication.remove(user);
	}


	while(System.currentTimeMillis() - startTime < TIME_LIMIT) {
		System.err.println("Thread Number: " + Thread.activeCount());
		executor.submit(new Runnable() {
			@Override
			public void run() {
				mApplication.acquireTokenSilentAsync(scopes, user, null, forceRefresh, new AuthenticationCallback() {
					@Override
					public void onSuccess(AuthenticationResult authenticationResult) {
						System.err.println("success");
					}

					@Override
					public void onError(MsalException exception) {
						System.err.println("error");
					}

					@Override
					public void onCancel() {
						System.err.println("cancel");
					}
				});
			}
		});
		Thread.sleep(5);
	}
	executor.shutdownNow();
	System.err.println( "All done." );
}
```

[msal_outofmemory_log.txt](https://github.com/AzureAD/microsoft-authentication-library-for-android/files/965219/msal_outofmemory_log.txt)

**Comment by heidijinxujia** [comment]
--------- beginning of crash
05-04 16:01:27.732 24198-24648/? E/AndroidRuntime: FATAL EXCEPTION: pool-1-thread-1
                                                   Process: com.microsoft.identity.client.sample, PID: 24198
                                                   java.lang.OutOfMemoryError: Failed to allocate a 81826734 byte allocation with 8314078 free bytes and 7MB until OOM
                                                       at java.lang.AbstractStringBuilder.enlargeBuffer(AbstractStringBuilder.java:95)
                                                       at java.lang.AbstractStringBuilder.append0(AbstractStringBuilder.java:146)
                                                       at java.lang.StringBuffer.append(StringBuffer.java:219)
                                                       at com.microsoft.identity.client.sample.MsalSampleApp$1.log(MsalSampleApp.java:50)
                                                       at com.microsoft.identity.client.Logger.log(Logger.java:216)
                                                       at com.microsoft.identity.client.Logger.verbose(Logger.java:178)
                                                       at com.microsoft.identity.client.HttpRequest.executeHttpSend(HttpRequest.java:197)
                                                       at com.microsoft.identity.client.HttpRequest.sendWithRetry(HttpRequest.java:151)
                                                       at com.microsoft.identity.client.HttpRequest.send(HttpRequest.java:132)
                                                       at com.microsoft.identity.client.HttpRequest.sendGet(HttpRequest.java:123)
                                                       at com.microsoft.identity.client.Oauth2Client.executeHttpRequest(Oauth2Client.java:140)
                                                       at com.microsoft.identity.client.Oauth2Client.discoverEndpoints(Oauth2Client.java:113)
                                                       at com.microsoft.identity.client.Authority.resolveEndpoints(Authority.java:154)
                                                       at com.microsoft.identity.client.BaseRequest.preTokenRequest(BaseRequest.java:59)
                                                       at com.microsoft.identity.client.SilentRequest.preTokenRequest(SilentRequest.java:75)
                                                       at com.microsoft.identity.client.BaseRequest$1.run(BaseRequest.java:105)
                                                       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
                                                       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
                                                       at java.lang.Thread.run(Thread.java:818)

**Comment by weijjia** [comment]
OK, then it's a problem that how sample is appending all the log messages. I'll leave it as it is for now and fix it after build.

**Comment by weijjia** [comment]
Fixed with PR #191
